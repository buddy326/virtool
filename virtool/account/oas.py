from typing import Union, Optional
from pydantic import BaseModel, constr, Field, root_validator, validator
from email_validator import validate_email, EmailSyntaxError
from virtool.users.utils import Permission
from virtool_core.models.enums import QuickAnalyzeWorkflow


def check_permissions(permissions: dict) -> dict:
    if any(not hasattr(Permission, key) for key in permissions):
        raise ValueError("One or more permissions is invalid")

    return permissions


class EditAccountSchema(BaseModel):
    """
    Fields for editing a user account.
    """

    email: constr(strip_whitespace=True)
    old_password: str
    password: str

    @root_validator
    def check_password(cls, values: Union[str, constr]):
        """
        Checks if old_password has also been input if a new password
        is provided.
        """
        old_password, password = values.get("old_password"), values.get("password")

        if password is not None and old_password is None:
            raise ValueError(
                "The old password needs to be given in order for the password to be changed"
            )

        return values

    @validator("email")
    def check_email(cls, email: Optional[str]) -> str:
        """
        Checks if the given email is valid.
        """
        try:
            validate_email(email)
        except EmailSyntaxError:
            raise ValueError("The format of the email is invalid")

        return email


class EditSettingsSchema(BaseModel):
    """
    Fields for editing a user account's settings.
    """

    show_ids: Optional[bool]
    skip_quick_analyze_dialog: Optional[bool]
    quick_analyze_workflow: Optional[QuickAnalyzeWorkflow]


class CreateKeysSchema(BaseModel):
    name: constr(strip_whitespace=True, min_length=1)
    permissions: dict = Field(default={})

    _ensure_permissions_is_valid = validator("permissions", allow_reuse=True)(
        check_permissions
    )


class EditKeySchema(BaseModel):
    permissions: dict

    _ensure_permissions_is_valid = validator("permissions", allow_reuse=True)(
        check_permissions
    )


class CreateLoginSchema(BaseModel):
    username: constr(min_length=1)
    password: constr(min_length=1)
    remember: bool = Field(default=False)


class ResetPasswordSchema(BaseModel):
    password: str
    reset_code: str
