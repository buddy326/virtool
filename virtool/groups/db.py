"""
Database utilities for groups.

"""

import asyncio
from asyncio import gather
from typing import List, Optional

from motor.motor_asyncio import AsyncIOMotorClientSession
from virtool_core.models.group import Group
from virtool_core.models.user import UserNested
from virtool.groups.pg import SQLGroup
from virtool.pg.utils import get_row, get_row_by_id

import virtool.users.db
import virtool.utils
from virtool.groups.utils import merge_group_permissions
from virtool.utils import base_processor


async def fetch_complete_group(mongo, pg, group_id: str | int) -> Optional[Group]:
    """
    Search Mongo and Postgres for a Group by its ID.

    :param mongo: The MongoDB client
    :param pg: PostgreSQL AsyncEngine object
    :param group_id: ID to search group by
    """

    if type(group_id) is str:
        return await fetch_complete_group_by_legacy_id(mongo, pg, group_id)

    pg_group = await get_row_by_id(pg, SQLGroup, group_id)

    if pg_group:
        return Group(
            **{
                **pg_group.to_dict(),
                "id": pg_group.legacy_id,
                "users": await fetch_group_users(mongo, pg_group.legacy_id),
            }
        )


async def fetch_complete_group_by_legacy_id(
    mongo, pg, group_id: str
) -> Optional[Group]:
    """
    Search Mongo and Postgres for a Group by its legacy mongo id

    :param mongo: The MongoDB client
    :param pg: PostgreSQL AsyncEngine object
    :param group_id: ID to search group by
    """
    pg_group, mongo_group, users = await gather(
        get_row(pg, SQLGroup, ("legacy_id", group_id)),
        mongo.groups.find_one({"_id": group_id}),
        fetch_group_users(mongo, group_id),
    )

    if pg_group:
        group = pg_group.to_dict()
        group["id"] = pg_group.legacy_id
    else:
        group = mongo_group

    if group:
        group["users"] = users
        return Group(**group)


async def get_merged_permissions(mongo, id_list: List[str]) -> dict:
    """
    Get the merged permissions that are inherited as a result of membership in the groups defined in `id_list`.

    :param mongo: The MongoDB client
    :param id_list: a list of group ids
    :return: the merged permissions

    """
    groups = await asyncio.shield(
        mongo.groups.find({"_id": {"$in": id_list}}, {"_id": False}).to_list(None)
    )
    return merge_group_permissions(groups)


async def update_member_users(
    mongo,
    group_id: str,
    remove: bool = False,
    session: Optional[AsyncIOMotorClientSession] = None,
):
    groups = await mongo.groups.find({}, session=session).to_list(None)

    async for user in mongo.users.find(
        {"groups": group_id},
        ["administrator", "groups", "permissions", "primary_group"],
        session=session,
    ):
        if remove:
            user["groups"].remove(group_id)

        update_dict = {}

        if remove:
            update_dict["$pull"] = {"groups": group_id}

            if user["primary_group"] == group_id:
                update_dict["$set"]["primary_group"] = ""

        if update_dict:
            await mongo.users.find_one_and_update(
                {"_id": user["_id"]},
                update_dict,
                projection=["groups", "permissions"],
                session=session,
            )

        permissions = merge_group_permissions(
            [group for group in groups if group["_id"] in user["groups"]]
        )

        await virtool.users.db.update_keys(
            mongo,
            user["administrator"],
            user["_id"],
            user["groups"],
            permissions,
            session=session,
        )


async def fetch_group_users(mongo, group_id: str) -> List[UserNested]:
    return [
        UserNested(**base_processor(user))
        async for user in mongo.users.find({"groups": group_id})
    ]
