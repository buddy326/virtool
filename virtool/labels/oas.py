from pydantic import BaseModel, constr, Field, validator

from virtool_core.models import normalize_hex_color


class CreateLabelSchema(BaseModel):
    """
    Label fields for creating a new label.
    """

    name: constr(strip_whitespace=True, min_length=1) = Field(
        description="unique name for the label document"
    )
    color: constr(strip_whitespace=True) = Field(default="#A0AEC0")
    description: constr(strip_whitespace=True) = Field(
        default="", description="description of the document"
    )

    _ensure_color_is_normalized: classmethod = validator("color", allow_reuse=True)(
        normalize_hex_color
    )


class EditLabelSchema(BaseModel):
    """
    Label fields for editing an existing label.
    """

    name: constr(strip_whitespace=True) = Field(
        description="name of the existing label document"
    )
    color: constr(strip_whitespace=True)
    description: constr(strip_whitespace=True) = Field(
        description="description of the existing label document"
    )

    _ensure_color_is_normalized: classmethod = validator("color", allow_reuse=True)(
        normalize_hex_color
    )
