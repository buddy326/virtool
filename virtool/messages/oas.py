from typing import Optional

from pydantic import BaseModel, validator
from virtool_core.models.enums import MessageColor
from virtool_core.models.validators import prevent_none


class CreateMessageRequest(BaseModel):
    color: MessageColor
    message: str


class UpdateMessageRequest(BaseModel):
    color: Optional[MessageColor]
    message: Optional[str]
    active: Optional[bool]

    _prevent_none = prevent_none('*')

    @validator("active")
    def active_must_be_false(cls, active: bool) -> bool:
        if active:
            raise ValueError("active can only be `False` when updating")
        return active
