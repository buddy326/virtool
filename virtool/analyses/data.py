from asyncio import gather, CancelledError
from datetime import datetime
from typing import List, Union, Tuple, Dict, Optional, Any
from logging import getLogger

from multidict import MultiDictProxy
from sqlalchemy.ext.asyncio import AsyncEngine
from virtool_core.models.analysis import AnalysisMinimal, Analysis

import virtool.samples.db
import virtool.analyses.format
from virtool.analyses.db import PROJECTION, processor
from virtool.analyses.files import create_analysis_file
from virtool.analyses.models import AnalysisFile
from virtool.analyses.utils import attach_analysis_files, find_nuvs_sequence_by_index
from virtool.api.utils import paginate
from virtool.blast.transform import AttachNuVsBLAST
from virtool.data.errors import (
    ResourceNotFoundError,
    ResourceNotModifiedError,
    ResourceError,
    ResourceConflictError,
)
from virtool.mongo.core import DB
import virtool.uploads.db
from virtool.mongo.transforms import apply_transforms
from virtool.mongo.utils import get_one_field
from virtool.pg.utils import delete_row, get_row_by_id
from virtool.samples.utils import get_sample_rights
from virtool.subtractions.db import AttachSubtractionTransform
from virtool.types import Document
from virtool.uploads.utils import naive_writer

from virtool.users.db import AttachUserTransform
from virtool.utils import run_in_thread, rm
from virtool.samples.db import recalculate_workflow_tags


logger = getLogger("analyses")


class AnalysisData:
    def __init__(self, db: DB, config, pg: AsyncEngine):
        self._db = db
        self._config = config
        self._pg = pg

    async def find(
        self, query: Union[Dict, MultiDictProxy[str]], client
    ) -> Tuple[dict, List[AnalysisMinimal]]:
        """
        List all analysis documents.

        :param query: the request query
        :param client: the client object

        :return: a list of all analysis documents
        """
        db_query = {}

        data = await paginate(
            self._db.analyses,
            db_query,
            query,
            projection=PROJECTION,
            sort=[("created_at", -1)],
        )

        per_document_can_read = await gather(
            *[
                virtool.samples.db.check_rights(
                    self._db, document["sample"]["id"], client, write=False
                )
                for document in data["documents"]
            ]
        )

        documents = [
            document
            for document, can_write in zip(data["documents"], per_document_can_read)
            if can_write
        ]

        documents = await apply_transforms(
            documents,
            [AttachUserTransform(self._db), AttachSubtractionTransform(self._db)],
        )

        for analysis in documents:
            if "updated_at" not in analysis:
                analysis.update({"updated_at": analysis["created_at"]})

        return data, [AnalysisMinimal(**document) for document in documents]

    async def get(
        self, analysis_id: str, if_modified_since: Union[datetime, None]
    ) -> Union[Tuple[Document, datetime], Tuple[Analysis, datetime]]:
        """
        Get a single analysis by its ID.

        :param analysis_id: the analysis ID
        :param if_modified_since: the date tbe document should have been last modified by

        :return: the analysis
        """
        document = await self._db.analyses.find_one(analysis_id)

        created_at = document["created_at"]

        if document is None:
            raise ResourceNotFoundError()

        if if_modified_since is not None:
            try:
                updated_at = document["updated_at"]
            except KeyError:
                updated_at = document["created_at"]

            if if_modified_since and if_modified_since == updated_at:
                raise ResourceNotModifiedError()

        document = await attach_analysis_files(self._pg, analysis_id, document)

        sample = await self._db.samples.find_one(
            {"_id": document["sample"]["id"]}, {"quality": False}
        )

        if not sample:
            raise ResourceError()

        if document["ready"]:
            document = await virtool.analyses.format.format_analysis(
                self._config, self._db, document
            )

        document = await processor(self._db, document)

        if document["workflow"] == "nuvs":
            document = await apply_transforms(
                document,
                [AttachNuVsBLAST(self._pg)],
            )

        if "index" not in document:
            return document, created_at

        return Analysis(**document), created_at

    async def has_right(self, analysis_id, client, right) -> bool:
        """
        Checks if the client has the `read` or `write` rights.

        :param analysis_id: the analysis ID
        :param client: the client object
        :param right: the right to check for
        """
        sample = await get_one_field(self._db.analyses, "sample", analysis_id)

        if sample is None:
            raise ResourceNotFoundError()

        sample = await self._db.samples.find_one(
            {"_id": sample["id"]},
            ["user", "group", "all_read", "group_read", "group_write", "all_write"],
        )

        if not sample:
            raise ResourceError()

        read, write = get_sample_rights(sample, client)

        if right == "read":
            return read

        if right == "write":
            return write

    async def delete(self, analysis_id: str, jobs_api_flag: int):
        """
        Delete a single analysis by its ID.

        :param analysis_id: the analysis ID
        :param jobs_api_flag: checks if the jobs_api is handling the request
        """
        document = await self._db.analyses.find_one(
            {"_id": analysis_id}, ["job", "ready", "sample"]
        )

        if not document:
            raise ResourceNotFoundError()

        sample_id = document["sample"]["id"]

        if jobs_api_flag:
            if document["ready"]:
                raise ResourceConflictError()
        else:
            if not document["ready"]:
                raise ResourceConflictError()

        await self._db.analyses.delete_one({"_id": analysis_id})

        path = self._config.data_path / "samples" / sample_id / "analysis" / analysis_id

        try:
            await run_in_thread(rm, path, True)
        except FileNotFoundError:
            pass

        await recalculate_workflow_tags(self._db, sample_id)

    async def upload(
        self, reader, analysis_id: str, analysis_format: str, name: str
    ) -> Union[int, dict]:
        """
        Uploads a new analysis result file.

        :param reader: the file reader
        :param analysis_id: the analysis ID
        :param analysis_format: the format of the analysis
        :param name: the name of the analysis file

        :return: the new analysis file
        """

        document = await self._db.analyses.find_one(analysis_id)

        if document is None:
            raise ResourceNotFoundError()

        analysis_file = await create_analysis_file(
            self._pg, analysis_id, analysis_format, name
        )

        upload_id = analysis_file["id"]

        analysis_file_path = (
            self._config.data_path / "analyses" / analysis_file["name_on_disk"]
        )

        try:
            size = await naive_writer(reader, analysis_file_path)
        except CancelledError:
            logger.debug(f"Analysis file upload aborted: {upload_id}")
            await delete_row(self._pg, upload_id, AnalysisFile)

            return 499

        analysis_file = await virtool.uploads.db.finalize(
            self._pg, size, upload_id, AnalysisFile
        )

        return analysis_file

    async def get_file(self, upload_id: int) -> str:
        """
        Get a file generated during the analysis.

        :param upload_id: the upload ID

        :return: the name on disk of the analysis file
        """

        analysis_file = await get_row_by_id(self._pg, AnalysisFile, upload_id)

        if not analysis_file:
            raise ResourceNotFoundError()

        return analysis_file.name_on_disk

    async def download_analysis(
        self, analysis_id: str, extension: str
    ) -> Tuple[bytes, str]:
        """
        Get an analysis to be downloaded in CSV or XSLX format.

        :param analysis_id: the analysis ID
        :param extension: the file extension

        :return: formatted file and file content type
        """
        document = await self._db.analyses.find_one(analysis_id)

        if not document:
            raise ResourceNotFoundError()

        if extension == "xlsx":
            formatted = await virtool.analyses.format.format_analysis_to_excel(
                self._config, self._db, document
            )
            content_type = (
                "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
            )
        else:
            formatted = await virtool.analyses.format.format_analysis_to_csv(
                self._config, self._db, document
            )
            content_type = "text/csv"

        return formatted, content_type

    async def blast(self, analysis_id: str, sequence_index: int) -> Optional[str]:
        """
        BLAST a contig sequence that is part of a NuVs result record.

        :param analysis_id: the analysis ID
        :param sequence_index: the sequence index

        :return: the nuvs sequence
        """
        document = await self._db.analyses.find_one(
            {"_id": analysis_id}, ["ready", "workflow", "results", "sample"]
        )

        if document["workflow"] != "nuvs":
            raise ResourceConflictError("Not a NuVs analysis")

        if not document["ready"]:
            raise ResourceConflictError("Analysis is still running")

        return find_nuvs_sequence_by_index(document, sequence_index)

    async def finalize(self, analysis_id: str, results: dict) -> Dict[str, Any]:
        """
        Sets the result for an analysis and marks it as ready.

        :param analysis_id: the analysis ID
        :param results: the analysis results

        :return: the processed analysis document
        """

        analysis_document = await self._db.analyses.find_one(
            {"_id": analysis_id}, ["ready"]
        )

        if not analysis_document:
            raise ResourceNotFoundError()

        if "ready" in analysis_document and analysis_document["ready"]:
            raise ResourceConflictError

        document = await self._db.analyses.find_one_and_update(
            {"_id": analysis_id}, {"$set": {"results": results, "ready": True}}
        )

        await recalculate_workflow_tags(self._db, document["sample"]["id"])
        await attach_analysis_files(self._pg, analysis_id, document)

        return await processor(self._db, document)
