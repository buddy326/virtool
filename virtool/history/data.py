from typing import List, Any

from aiohttp.web_app import Application
from virtool_core.models.history import HistorySearchResult, History

import virtool
from virtool.data.errors import ResourceNotFoundError, ResourceConflictError
from virtool.errors import DatabaseError
from virtool.history.db import LIST_PROJECTION
from virtool.mongo.core import DB


class HistoryData:
    def __init__(self, db: DB):
        self._db = db

    async def find(self, req_query: Any) -> HistorySearchResult:
        """
        List all change documents.
        :param req_query: the request query
        :return: a list of all documents
        """

        documents = await virtool.history.db.find(db=self._db, req_query=req_query)

        return HistorySearchResult(**documents)

    async def get(self, app: Application, change_id: str) -> History:
        """
        Get a document given its ID.
        :param app: the application object
        :param change_id: the ID of the document to delete
        """

        document = await virtool.history.db.get(app, change_id)

        if not document:
            raise ResourceNotFoundError()

        return History(**document)

    async def delete(self, app: Application, change_id: str):
        """
        Delete a document given its ID.
        :param app: the application object
        :param change_id: the ID of the document to delete
        """
        document = await self._db.history.find_one(change_id, ["reference"])

        if not document:
            raise ResourceNotFoundError()

        try:
            await virtool.history.db.revert(app, change_id)
        except DatabaseError:
            raise ResourceConflictError()
