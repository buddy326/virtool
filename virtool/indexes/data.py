import asyncio
import json
import logging
from pathlib import Path
from typing import List, Union, Optional

from multidict import MultiDictProxy
from sqlalchemy.exc import IntegrityError
from sqlalchemy.ext.asyncio import AsyncEngine
from virtool.api.utils import compose_regex_query, paginate
from virtool.history.db import LIST_PROJECTION
from virtool_core.models.history import HistorySearchResult
from virtool_core.models.index import IndexMinimal, IndexSearchResult, Index
from virtool_core.models.reference import ReferenceNested

from virtool.api.custom_json import CustomEncoder
from virtool.config import Config
from virtool.data.errors import ResourceNotFoundError, ResourceConflictError
from virtool.indexes.db import (
    FILES,
)
from virtool.indexes.models import IndexFile, IndexType

import virtool.indexes.db
from virtool.indexes.files import create_index_file
from virtool.indexes.utils import join_index_path
from virtool.mongo.core import DB
from virtool.pg.utils import delete_row, get_rows
from virtool.uploads.utils import naive_writer
from virtool.utils import run_in_thread, compress_json_with_gzip
import virtool.uploads.db

logger = logging.getLogger("indexes")


class IndexData:
    def __init__(self, db: DB, config: Config, pg: AsyncEngine):
        self._db = db
        self._config = config
        self._pg = pg

    async def find(
        self, ready: bool, query: MultiDictProxy[str]
    ) -> Union[List[IndexMinimal], IndexSearchResult]:
        """
        List all indexes.

        :param ready: the request object
        :param query: the request query object
        :return: a list of all index documents
        """
        if not ready:
            data = await virtool.indexes.db.find(self._db, query)
            return IndexSearchResult(**data)

        pipeline = [
            {"$match": {"ready": True}},
            {"$sort": {"version": -1}},
            {
                "$group": {
                    "_id": "$reference.id",
                    "index": {"$first": "$_id"},
                    "version": {"$first": "$version"},
                    "change_count": {"$first": "$change_count"},
                    "created_at": {"$first": "$created_at"},
                    "has_files": {"$first": "$has_files"},
                    "modified_otu_count": {"$first": "$modified_otu_count"},
                    "user": {"$first": "$user"},
                    "ready": {"$first": "$ready"},
                    "job": {"$first": "$job"},
                }
            },
        ]

        ready_indexes = []

        async for agg in self._db.indexes.aggregate(pipeline):
            user = await self._db.users.find_one(agg["user"]["id"])
            ready_indexes.append(
                {
                    "id": agg["index"],
                    "version": agg["version"],
                    "reference": {
                        "id": agg["_id"],
                    },
                    "change_count": agg["change_count"],
                    "created_at": agg["created_at"],
                    "has_files": agg["has_files"],
                    "modified_otu_count": agg["modified_otu_count"],
                    "user": {
                        "id": user["_id"],
                        "handle": user["handle"],
                        "administrator": user["administrator"],
                    },
                    "ready": agg["ready"],
                    "job": agg["job"],
                }
            )

        return [IndexMinimal(**index) for index in ready_indexes]

    async def get(self, index_id: str) -> Index:
        """
        Get a single index by its ID.

        :param index_id: the index ID
        :return: the index
        """
        document = await self._db.indexes.find_one(index_id)

        if not document:
            raise ResourceNotFoundError()

        contributors, otus = await asyncio.gather(
            virtool.history.db.get_contributors(self._db, {"index.id": index_id}),
            virtool.indexes.db.get_otus(self._db, index_id),
        )

        document.update(
            {
                "change_count": sum(v["change_count"] for v in otus),
                "contributors": contributors,
                "otus": otus,
            }
        )

        document = await virtool.indexes.db.attach_files(
            self._pg, self._config.base_url, document
        )

        document = await virtool.indexes.db.processor(self._db, document)

        return Index(**document)

    async def get_reference(self, index_id: str) -> ReferenceNested:
        """
        Get a reference associated with an index.

        :param index_id: the index ID
        :return: the reference
        """
        index = await self._db.indexes.find_one(index_id)

        if index is None:
            raise ResourceNotFoundError()

        reference = index["reference"]

        return ReferenceNested(**reference)

    async def get_json_path(self, index_id: str) -> Path:
        """
        Get the json path needed for a complete compressed JSON
        representation of the index OTUs.

        :param index_id: the index ID
        :return: the json path
        """
        index = await self._db.indexes.find_one(index_id)

        if index is None:
            raise ResourceNotFoundError()

        ref_id = index["reference"]["id"]

        json_path = (
            join_index_path(self._config.data_path, ref_id, index_id) / "otus.json.gz"
        )

        if not json_path.exists():
            patched_otus = await virtool.indexes.db.get_patched_otus(
                self._db, self._config, index["manifest"]
            )

            json_string = json.dumps(patched_otus, cls=CustomEncoder)

            await run_in_thread(compress_json_with_gzip, json_string, json_path)

        return json_path

    async def upload_file(
        self, reference_id: str, index_id: str, file_type: str, name: str, multipart
    ) -> Optional[IndexFile]:
        """
        Uploads a new index file.

        :param reference_id: the reference ID
        :param index_id: the index ID
        :param file_type: the type of the file to upload
        :param name: the name of the new file
        :param multipart: the file reader
        :return: the index file
        """
        try:
            index_file = await create_index_file(self._pg, index_id, file_type, name)
        except IntegrityError:
            raise ResourceConflictError()

        upload_id = index_file["id"]

        path = join_index_path(self._config.data_path, reference_id, index_id) / name

        try:
            size = await naive_writer(await multipart(), path)
        except asyncio.CancelledError:
            logger.debug("Index file upload aborted: %s", upload_id)
            await delete_row(self._pg, upload_id, IndexFile)
            return None

        index_file = await virtool.uploads.db.finalize(
            self._pg, size, upload_id, IndexFile
        )

        return IndexFile(**index_file)

    async def finalize(self, ref_id: str, index_id: str) -> Index:
        """
        Finalize an index document.

        :param ref_id: the reference ID
        :param index_id: the index ID
        :return: the finalized Index
        """
        reference = await self._db.references.find_one(ref_id)

        if reference is None:
            raise ResourceNotFoundError()

        rows = await get_rows(self._pg, IndexFile, "index", index_id)

        results = {f.name: f.type for f in rows}

        if IndexType.fasta not in results.values():
            raise ResourceConflictError(
                "A FASTA file must be uploaded in order to finalize index"
            )

        if reference["data_type"] == "genome":
            required_files = [f for f in FILES if f != "reference.json.gz"]

            if missing_files := [f for f in required_files if f not in results]:
                raise ResourceConflictError(
                    f"Reference requires that all Bowtie2 index files have been uploaded. "
                    f"Missing files: {', '.join(missing_files)}"
                )

        document = await virtool.indexes.db.finalize(
            self._db, self._pg, self._config.base_url, ref_id, index_id
        )

        document = await self.get(index_id)

        return document

    async def find_changes(
        self, index_id: str, req_query: MultiDictProxy[str]
    ) -> HistorySearchResult:
        """
        Find the virus changes that are included in a given index build.

        :param index_id: the index ID
        :param req_query: the request query object
        :return: the changes
        """
        if not await self._db.indexes.count_documents({"_id": index_id}):
            raise ResourceNotFoundError()

        db_query = {"index.id": index_id}

        term = req_query.get("term")

        if term:
            db_query.update(compose_regex_query(term, ["otu.name", "user.id"]))

        data = await paginate(
            self._db.history,
            db_query,
            req_query,
            sort=[("otu.name", 1), ("otu.version", -1)],
            projection=LIST_PROJECTION,
            reverse=True,
        )

        return HistorySearchResult(**data)

    async def delete(self, index_id: str):
        """
        Delete an index given it's id.

        :param index_id: the ID of the index to delete
        """

        delete_result = await self._db.indexes.delete_one({"_id": index_id})

        if delete_result.deleted_count != 1:
            # Document could not be deleted.
            raise ResourceNotFoundError()

        query = {
            "_id": {
                "$in": await self._db.history.distinct("_id", {"index.id": index_id})
            }
        }

        await self._db.history.update_many(
            query, {"$set": {"index": {"id": "unbuilt", "version": "unbuilt"}}}
        )
