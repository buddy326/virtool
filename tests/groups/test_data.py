import pytest
from sqlalchemy.ext.asyncio import AsyncEngine

from virtool.authorization.client import AuthorizationClient
from virtool.data.errors import ResourceConflictError, ResourceNotFoundError
from virtool.groups.data import GroupsData
from virtool.groups.oas import UpdateGroupRequest
from virtool.groups.pg import SQLGroup
from virtool.mongo.core import Mongo
from virtool.pg.utils import get_row_by_id


@pytest.fixture
def groups_data(authorization_client, mongo, pg):
    return GroupsData(authorization_client, mongo, pg)


async def test_create(
    groups_data: GroupsData,
    mongo: Mongo,
    pg: AsyncEngine,
    snapshot,
):
    group = await groups_data.create("test")
    row = await get_row_by_id(pg, SQLGroup, 1)
    doc = await mongo.groups.find_one()

    assert doc == snapshot(name="mongo")
    assert row == snapshot(name="pg")

    assert doc["name"] == group.name == row.name
    assert doc["permissions"] == group.permissions == row.permissions
    assert doc["_id"] == group.id == row.legacy_id


class TestGet:
    async def test_get(self, groups_data: GroupsData, fake2, snapshot):
        """
        Ensure the correct group is returned when passed an postgres integer ID
        """
        await fake2.groups.create()

        assert await groups_data.get(1) == snapshot

    async def test_legacy_id(self, groups_data: GroupsData, fake2, snapshot):
        """
        Ensure the correct group is returned when passed a legacy mongo id
        """
        group = await fake2.groups.create()

        assert await groups_data.get(group.id) == snapshot

    async def test_user(self, groups_data: GroupsData, fake2, snapshot):
        """
        Ensure that users are correctly attached to the returned groups
        """
        group = await fake2.groups.create()

        await fake2.users.create(groups=[group])

        assert await groups_data.get(1) == snapshot

    @pytest.mark.parametrize("group_id", ["group_dne", 0xBEEF])
    async def test_group_dne(self, groups_data: GroupsData, group_id: str | int):
        """
        Ensure the correct exception is raised when the group does not exist
        using either a postgres or mongo id
        """
        with pytest.raises(ResourceNotFoundError):
            await groups_data.get(group_id)


async def test_create_duplicate(groups_data: GroupsData):
    group = await groups_data.create("test")

    with pytest.raises(ResourceConflictError):
        await groups_data.create(group.name)


async def test_update_name(
    authorization_client: AuthorizationClient,
    groups_data: GroupsData,
    mongo: Mongo,
    pg: AsyncEngine,
    snapshot,
):
    group = await groups_data.create("Test")

    document = await mongo.groups.find_one()
    row = await get_row_by_id(pg, SQLGroup, 1)

    assert document["name"] == row.name == "Test" == group.name

    group = await groups_data.update(group.id, UpdateGroupRequest(name="Renamed"))

    document = await mongo.groups.find_one()
    row = await get_row_by_id(pg, SQLGroup, 1)

    assert document["name"] == row.name == "Renamed" == group.name


async def test_update_permissions(
    authorization_client: AuthorizationClient,
    groups_data: GroupsData,
    mongo: Mongo,
    pg: AsyncEngine,
    snapshot,
):
    group = await groups_data.create("Test")

    group = await groups_data.update(
        group.id,
        UpdateGroupRequest(
            **{"permissions": {"create_sample": True, "modify_subtraction": True}}
        ),
    )

    doc = await mongo.groups.find_one()
    row = await get_row_by_id(pg, SQLGroup, 1)

    assert (
        group.permissions
        == row.permissions
        == doc["permissions"]
        == snapshot(name="mongo_added")
    )

    group = await groups_data.update(
        group.id, UpdateGroupRequest(**{"permissions": {"create_sample": False}})
    )

    doc = await mongo.groups.find_one()
    row = await get_row_by_id(pg, SQLGroup, 1)

    assert (
        group.permissions
        == row.permissions
        == doc["permissions"]
        == snapshot(name="mongo_removed")
    )


async def test_delete(authorization_client, groups_data, mongo, pg, snapshot):
    group = await groups_data.create("Test")

    await groups_data.delete(group.id)

    with pytest.raises(ResourceNotFoundError):
        await groups_data.get(group.id)

    assert await mongo.groups.count_documents({}) == 0
    assert await get_row_by_id(pg, SQLGroup, 1) is None


async def test_delete_dne(groups_data: GroupsData, snapshot):
    with pytest.raises(ResourceNotFoundError):
        await groups_data.delete("foobar")
