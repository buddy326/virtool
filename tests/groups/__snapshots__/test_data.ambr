# name: TestGet.test_get
  <class 'dict'> {
    'id': 'bf1b993c',
    'name': 'musicians',
    'permissions': <class 'dict'> {
      'cancel_job': False,
      'create_ref': False,
      'create_sample': False,
      'modify_hmm': False,
      'modify_subtraction': False,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    },
    'users': <class 'list'> [
    ],
  }
---
# name: TestGet.test_legacy_id
  <class 'dict'> {
    'id': 'bf1b993c',
    'name': 'musicians',
    'permissions': <class 'dict'> {
      'cancel_job': False,
      'create_ref': False,
      'create_sample': False,
      'modify_hmm': False,
      'modify_subtraction': False,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    },
    'users': <class 'list'> [
    ],
  }
---
# name: TestGet.test_user
  <class 'dict'> {
    'id': 'bf1b993c',
    'name': 'musicians',
    'permissions': <class 'dict'> {
      'cancel_job': False,
      'create_ref': False,
      'create_sample': False,
      'modify_hmm': False,
      'modify_subtraction': False,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    },
    'users': <class 'list'> [
      <class 'dict'> {
        'administrator': False,
        'handle': 'laura56',
        'id': 'fb085f7f',
      },
    ],
  }
---
# name: test_create[mongo]
  <class 'dict'> {
    '_id': 'bf1b993c',
    'name': 'test',
    'permissions': <class 'dict'> {
      'cancel_job': False,
      'create_ref': False,
      'create_sample': False,
      'modify_hmm': False,
      'modify_subtraction': False,
      'remove_file': False,
      'remove_job': False,
      'upload_file': False,
    },
  }
---
# name: test_create[pg]
  <SQLGroup(id=1, legacy_id=bf1b993c, name=test, permissions={'cancel_job': False, 'create_ref': False, 'modify_hmm': False, 'remove_job': False, 'remove_file': False, 'upload_file': False, 'create_sample': False, 'modify_subtraction': False})>
---
# name: test_update_permissions[mongo_added]
  <class 'dict'> {
    'cancel_job': False,
    'create_ref': False,
    'create_sample': True,
    'modify_hmm': False,
    'modify_subtraction': True,
    'remove_file': False,
    'remove_job': False,
    'upload_file': False,
  }
---
# name: test_update_permissions[mongo_removed]
  <class 'dict'> {
    'cancel_job': False,
    'create_ref': False,
    'create_sample': False,
    'modify_hmm': False,
    'modify_subtraction': True,
    'remove_file': False,
    'remove_job': False,
    'upload_file': False,
  }
---
